[JukeBoxK] iP

JukeBoxK opened the pull request:
# Alfred Pro
> "Your mind is for working, not storing." - Barry Allen ([source](https://www.google.com))

Alfred Pro helps you to remember yours tasks. It's
- Command-based
- Easy to understand
- ***Extremely*** **Fast** to use

Do these to get started:
1. Download the ~~file~~ app
2. Open the ~~file~~ app
3. Done!
4. Start adding your task now 🧇 

And it costs **NOTHING**!!

Features:
- [X] Managing tasks
- [X] Managing deadlines
- [X] Managing events
- [ ] Scheduler (coming soon)

If you are a programmer, you can even use it to code. Here's a `print` method:

```ruby
public void printHelloWord(){
    System.out.println("Hello World");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

JukeBoxK committed:
Level-0: Renamed the chatbot. and added greet and exit function

JukeBoxK committed:
Level-1: Added echo function and customized the personality of the chatbox

JukeBoxK committed:
Level-2: Added the ability to store text and show those stored text.

JukeBoxK committed:
Level-3: Added the ability to mark and unmark tasks.

JukeBoxK committed:
Level-4: Added 3 different type of tasks: Todos, Deadliens and Events

JukeBoxK committed:
A-TextUiTesting: Added Some test cases for semi automated CLI testing.

JukeBoxK committed:
Level-5: The chatbot can now handle common input error

JukeBoxK committed:
Level-6: Added a delete function to remove task from list.

JukeBoxK committed:
A-Enums: Change the command to enums.

JukeBoxK committed:
Add separate files for the Task class and its subclasses

The classes and functions are all written into the same file.

This makes it hard to maintain the clarity and understandbility of the code.

Separating the classes into different file helps to modularize the code and make them easier
to understand.

Let's start by creating separate files for the Task class and its subclasses

JukeBoxK committed:
Remove the corresponding code to the Task class and its subclasses

JukeBoxK committed:
Separate OutputMessage and TaskList as a standalone class

JukeBoxK committed:
Remove corresponding code to OutputMessage and TaskList from the original file

JukeBoxK committed:
Make minor changes to test files to reflect the change in logo

JukeBoxK committed:
Make a class for FileHandler and Command

The code for command is in the same file as main.

This makes it to maintain as the number of commands increases.

Let's put it in a separate file to fix this issue.

This makes it easier to maintian the code base.

A file for FileHandler is also created to handle the file system of the app.

JukeBoxK committed:
Add the task list saving feature

JukeBoxK committed:
Fix the bug of incorrecly displaying marked task as unmarked from the stored file

JukeBoxK committed:
Clean up the output message for app

JukeBoxK committed:
Add the gitignore to ignore the data folder for the storage of task lists

JukeBoxK committed:
Merge branch 'branch-level-7'

JukeBoxK committed:
Add Parser class for parsing string input and Command class and its subclasses

Command class does parsing pass the comaand to the Tasklist to handle it.

This is not consistent with the class definition and makes it hard to add new
commands to the app.

Let's create a new Parse class that does the parsing. Additionally, lets create
an abstract Command class that can run a function by itself and make all the
previous commands into individual classes that extends of the Command class.

This allows for better abstraction and makes the code more extensible.

Let's also add the Local Date time capability into the parser.

JukeBoxK committed:
Add date and time recognition for tasks that take in date and time as input

JukeBoxK committed:
Modify the input format and output format for date and time data to be different

JukeBoxK committed:
Merge branch 'branch-Level-8'

JukeBoxK committed:
Modify the code base to be more OOP oriented

The code strucutre of Duke.java is adheres a lot more to functional programming.
With that said, the other parts of the code base follow OOP paradigm more.

As a result, there are inconsistencies in the main paradigm used by the code.

Let's restructure Duke.java to follow a more OOP approach. Additionally, let's rename
FileHandler to Storage and OutputMessage to Ui.

This can help to improve the consistency of the code base and increase its understandability.

JukeBoxK committed:
Integrate the Parser and Ui more effectively by outputing informative error message

JukeBoxK committed:
Merge branch 'branch-A-MoreOOP'

JukeBoxK committed:
Divide the classes into different packages

All the class files are in the same file, classes with varying behaviour
and type are mixed into the same place.

This makes the codebase hard to manage as it grows larger.

To solve this issue, let's divide the all the class files into different
packages, by grouping similar classes into the same package.

This way, the code base becomes clearer and more understandable.

JukeBoxK committed:
Merge branch 'add-gradle-support'

JukeBoxK committed:
Add Gradle support to the program

JukeBoxK committed:
Use Junit library to test a Parser, and a Storage method

The program is tested manually and tested with a semi-automatic CLI
testing method which incorporate input/output redirection technique.

Manual testing can take a lot of time and CLI testing is limited in that
it cannot test each method separately, introducing the risk of missing
some bugs.

Let's use Junit library to unit test a few of the non-trivial method. We
first start with testing the crucial, parse method of the Parser class,
and the non trivial method of saving the file with each item added into
the list.

Using Junit can help to isolate and locate bugs in individual methods
more effectively. Additionally it can help to automate the test process,
saving a lot of time.

JukeBoxK committed:
Build a Jar Package using the shadowJar task from Gradle

JukeBoxK committed:
Start adding JavaDoc

JukeBoxK committed:
Start editing to follow the coding standard

JukeBoxK committed:
Start implementing task finding function

JukeBoxK committed:
Add Javadoc documentation to some classes

Documentation added to Duke, Storage, TaskList and Ui class

JukeBoxK committed:
Add JavaDoc to the remaining classes

The Documentation has been added to the Parser class, Command class and
its subclasses and the Task class and its subclasses.

JukeBoxK committed:
Modify the code to fit the Java coding standard

JukeBoxK committed:
Add the find function to the program

JukeBoxK committed:
Remove unnecessary comment

JukeBoxK committed:
Remove unnecessary comment

JukeBoxK committed:
Merge branch 'branch-A-JavaDoc'

JukeBoxK committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/command/AddDeadlineCommand.java
#	src/main/java/duke/command/AddEventCommand.java
#	src/main/java/duke/command/AddTodoCommand.java
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteTaskCommand.java
#	src/main/java/duke/command/ListTaskCommand.java
#	src/main/java/duke/command/NoActionCommand.java
#	src/main/java/duke/command/ToggleMarkTaskCommand.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/util/Parser.java
#	src/main/java/duke/util/Storage.java
#	src/main/java/duke/util/TaskList.java
#	src/main/java/duke/util/Ui.java

JukeBoxK committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/util/Parser.java
#	src/main/java/duke/util/TaskList.java
#	src/main/java/duke/util/Ui.java

tohlh commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,146 @@
+                } else {
+//                    System.out.println("Failed to create the directory.");
+                    return;
+                }
```
The `if`-block seems a bit redundant.

```suggestion
            if (!dir.exists()) {
                if (!dir.mkdirs()) {
                    return;
                }
```

tohlh commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,146 @@
+            }
+        }catch (FileNotFoundException e){
+            System.out.println("FILE NOT FOUND");
+        }
```
This seems to violate the coding styles for not adding whitespaces?

```suggestion
        try {
            Scanner s = new Scanner(f);
            while (s.hasNext()) {
                String str = s.nextLine();
                listStates.add(str);
                Parser.initializeTask(str,temp);
            }
        } catch (FileNotFoundException e){
            System.out.println("FILE NOT FOUND");
        }
```

tohlh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,53 @@
+            }while(!command.getType().equals(Parser.Cmd.bye));
+        }catch(IllegalArgumentException e) {
+            Ui.informInvalidCommand();
+        }
```
This seems to violate coding styles, might want to consider adding whitespaces?

```suggestion
        try {
            Command command;
            do {
                command = this.parser.parse(scanner.nextLine());
                command.run(this.taskList);
            } while(!command.getType().equals(Parser.Cmd.bye));
        } catch (IllegalArgumentException e) {
            Ui.informInvalidCommand();
        }
```

tohlh commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,122 @@
+            default:
+                System.out.println("PROBLEM encountered with the saved data while loading");
+                break;
+        }
```
The indentation seems to violate coding styles?

```suggestion
        switch (data[0]) {
        case "T":
            task = new Todo(data[2]);
            break;
        case "D":
            LocalDateTime date = LocalDateTime.parse(data[3], DateTimeFormatter.ofPattern("yyyy/MM/dd HHmm"));
            task = new Deadline(data[2], date);
            break;
        case "E":
            LocalDateTime fromDate = LocalDateTime.parse(data[3], DateTimeFormatter.ofPattern("yyyy/MM/dd HHmm"));
            LocalDateTime toDate = LocalDateTime.parse(data[4], DateTimeFormatter.ofPattern("yyyy/MM/dd HHmm"));
            task = new Event(data[2], fromDate, toDate);
            break;
        default:
            System.out.println("PROBLEM encountered with the saved data while loading");
            break;
        }
```

tohlh commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,146 @@
+ */
+public class Storage {
+    private ArrayList<String> listStates= new ArrayList<>();
+    private static String listFilePath = "./data/duke.txt";
```
Should make this a constant? Or maybe allow the user to assign a file path from the constructor?

tohlh commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,122 @@
+        String[] cmds = input.split(" ", 2);
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy/MM/dd HHmm");
+        try {
+            switch (Cmd.valueOf(cmds[0])) {
+                case bye:
+                    command = new ByeCommand(Cmd.bye);
+                    break;
+                case list:
+                    command = new ListTaskCommand(Cmd.list);
+                    break;
+                case mark:
+                    command = new ToggleMarkTaskCommand(Cmd.mark, Integer.parseInt(cmds[1]));
+                    break;
+                case unmark:
+                    command = new ToggleMarkTaskCommand(Cmd.unmark, Integer.parseInt(cmds[1]));
+                    break;
+                case todo:
+                    command = new AddTodoCommand(Cmd.todo, cmds[1]);
+                    break;
+                case deadline:
+                    String[] deadlineData = cmds[1].split(" /by ", 2);
+                    LocalDateTime date = LocalDateTime.parse(deadlineData[1], formatter);
+                    command = new AddDeadlineCommand(Cmd.deadline, deadlineData[0], date);
+                    break;
+                case event:
+                    String[] eventData = cmds[1].split(" /from ", 2);
+                    String[] eventData2 = eventData[1].split(" /to ", 2);
+                    LocalDateTime fromDate = LocalDateTime.parse(eventData2[0], formatter);
+                    LocalDateTime toDate = LocalDateTime.parse(eventData2[1], formatter);
+                    command = new AddEventCommand(Cmd.event, eventData[0], fromDate, toDate);
+                    break;
+                case find:
+                    command = new FindCommand(Cmd.find, cmds[1]);
+                    break;
+                case delete:
+                    command = new DeleteTaskCommand(Cmd.delete, Integer.parseInt(cmds[1]));
+                    break;
+                default:
+                    break;
```
The indentation seem to violate the coding styles?

tohlh commented on src/main/java/duke/command/ToggleMarkTaskCommand.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Represents the Command of marking/unmarking a task in a list.
+ */
+public class ToggleMarkTaskCommand extends Command{
```
Should class names be nouns? Maybe something like MarkTaskCommand.

tohlh commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Represents the Command of adding a new deadline to a task list.
+ */
+public class AddDeadlineCommand extends Command {
```
Should class names be nouns? Maybe can rename to something like `DeadlineCommand` here?

tohlh commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Represents the Command of adding a new event to a task list.
+ */
+public class AddEventCommand extends Command {
```
Should class names be nouns? Maybe can rename to something like `EventCommand` here?

tohlh commented on src/main/java/duke/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Represents the Command of adding a new todo to a task list.
+ */
+public class AddTodoCommand extends Command {
```
Should class names be nouns? Maybe can rename to something like `TodoCommand` here?

tohlh commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,46 @@
+package duke.command;
+
+import duke.util.Parser;
+import duke.util.TaskList;
+import duke.task.Event;
```
Should these be reordered in lexicographical order?

tohlh commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,122 @@
+package duke.util;
+
+import duke.command.*;
```
Should wildcard import be used?

tohlh reviewed (commented):
Overall, other than some coding standard issues, everything else LGTM!

Great job!!

tohlh commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,42 @@
+    private String description;
+    private LocalDateTime date;
+
+    /**
```
Good example of using javadocs format correctly!

JukeBoxK committed:
Implement Checkstyle

JukeBoxK committed:
Merge branch 'branch-A-CheckStyle'

Kaya3842 commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,115 @@
+ * Handles the operation of modifying the lists.
+ */
+public class TaskList {
+    private ArrayList<Task> listItems;
```
Should this be renamed to itemList?

Kaya3842 commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @param task the given task to be marked.
+     */
+    public static void informListMarked(Task task) {
```
Should this method be renamed as informTaskMarked? Because we are marking the tasks in the list not marking the list.

Kaya3842 commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @param task the given task to be unmarked.
+     */
+    public static void informListUnmarked(Task task) {
```
Same rename suggestion as above.
.

Kaya3842 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+            do {
+                command = this.parser.parse(scanner.nextLine());
+                command.run(this.taskList);
+            } while(!command.getType().equals(Parser.Cmd.bye));
```
Very small coding standard violation here (missing whitespace after while).

Kaya3842 commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,160 @@
+    private String logo = "\t  __   __    ____  ____  ____  ____\n"
+            + "\t / _\\ (  )  (  __)(  _ \\(  __)(    \\\n"
+            + "\t/    \\/ (_/\\ ) _)  )   / ) _)  ) D (\n"
+            + "\t\\_/\\_/\\____/(__)  (__\\_)(____)(____/\n";
```
Nice logo design!

Kaya3842 commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,160 @@
+                + "\t\tbye: to leave the program\n"
+                + "\t\tfind <keyword>: to find tasks with description that contains the keyword";
+        printMessageWithLines(message);
+    }
```
Isn't it better to give the correct format for only the wrong input that the user is typing? Giving them a whole list at once seems a bit overwhelming.

Kaya3842 reviewed (commented):
Great use of abstraction with lots of small classes. Also, you even divided into smaller packages for some set of classes. It makes the code very easy to read and follow. Some naming choices might be confusing. You can reconsider their names later.

JukeBoxK committed:
Set up JavaFX and get started with GUI

JukeBoxK committed:
Format the GUI

JukeBoxK committed:
Merge branch 'branch-Level-10'

JukeBoxK committed:
Add assertions into Parser and TaskList

The initialize method in Parser assumes that the input string to produce
the type of tasks is going to be just one single string. The mark,
unmark, and delete method of the TaskList assumes that the index input
are going to be greater than one.

There are just implicit assumption and are not checked while running
the program. As such, it is uncertain whether if these assumptions
hold true.

Using asserts statements make these assumptions explicit and they are
checked while the program is compiled.

Let's add some assert statements regarding this into the Parser and
TaskList class.

JukeBoxK committed:
Raise the code quality of the codes in util package

Util classes have complex methods due to their nature and as a result,
there are a number of code quality violations.

Common code quality violations include unused code fragments and methods
that have more than one layer of abstractions.

Resolving these code quality issues can help to make the code more
readable and understandable.

Lets remove all the unused code fragments and use additional method to
reduce the layer of abstractions in related methods.

The code in Task package and Command package is fine since, they all
have short LOC and implementation.

JukeBoxK committed:
Revert "Add assertions into Parser and TaskList"

This reverts commit ddbd775fc0a42398e9fd7d49eb15eb0e907405e8.

JukeBoxK committed:
Revert "Revert "Add assertions into Parser and TaskList""

This reverts commit b24e95af6743d624285746e7368f75077974b3d0.

JukeBoxK committed:
Merge pull request #1 from JukeBoxK/branch-A-Assertions

Add assertions into Parser and TaskList

JukeBoxK committed:
Merge branch 'master' into branch-A-CodeQuality

JukeBoxK committed:
Merge pull request #2 from JukeBoxK/branch-A-CodeQuality

Raise the code quality of the codes in util package

JukeBoxK committed:
Add view schedule function

Date values are being displayed for deadlines and events. But there is
no utilization for these values.

Let's make a view schedule function to find deadlines and events that
are specifically on the input date.

This can help to make the currently stored data more useful for the
user.

JukeBoxK committed:
Add Help Command

JukeBoxK committed:
Add error handling for index out of bound

JukeBoxK committed:
Add a user guide

JukeBoxK committed:
Improve code readability
